src/ast/writing/ast_mapping.ts

```diff
@@ -191,333 +221,393 @@ class LiteralWriter implements ASTNodeWriter {
-        const trueBody = writer.write(node.vTrueBody, fragments);
+function wrapCompoundStatement(
+    node: IfStatement | WhileStatement | ForStatement,
+    desc: SrcDesc
+): SrcDesc {
+    const last = desc[desc.length - 1];
+    if (last !== ";") {
+        return [[node, desc]];
+    }
 
-        if (node.vFalseBody) {
-            const falseBody = writer.write(node.vFalseBody, fragments);
+    return [[node, desc.slice(0, -1)], ";"];
+}
 
-            return `if (${condition}) ${trueBody} else ${falseBody}`;
-        }
+/**
+ * Compound statemetns don't have their own semicolons. However if a
+ * child has a semi-colon, we must make sure to exclude it from our soruce map.
+ */
+abstract class CompoundStatementWriter<
+    T extends IfStatement | ForStatement | WhileStatement
+> extends SimpleStatementWriter<T> {
+    write(node: T, writer: ASTWriter): SrcDesc {
+        const stmtDesc = this.writeInt(node, writer);
+        pushSemicolonsDown(stmtDesc);
 
-        return `if (${condition}) ${trueBody}`;
+        return wrapCompoundStatement(node, stmtDesc);
     }
 }
 
-class ForStatementWriter implements ASTNodeWriter {
-    write(node: ForStatement, writer: ASTWriter, fragments: Map<ASTNode, string>): string {
-        const body = writer.write(node.vBody, fragments);
-
-        const header = [
-            node.vInitializationExpression
-                ? writer.write(node.vInitializationExpression, fragments)
-                : "",
-            node.vCondition ? writer.write(node.vCondition, fragments) : "",
-            node.vLoopExpression ? writer.write(node.vLoopExpression, fragments) : ""
-        ];
+class IfStatementWriter extends CompoundStatementWriter<IfStatement> {
+    writeInt(node: IfStatement, writer: ASTWriter): SrcDesc {
+        if (node.vFalseBody) {
+            return writer.desc(
+                "if (",
+                node.vCondition,
+                ") ",
+                node.vTrueBody,
+                " else ",
+                node.vFalseBody
+            );
+        }
 
-        return "for (" + header.join("; ") + ") " + body;
+        return writer.desc("if (", node.vCondition, ") ", node.vTrueBody);
     }
 }
 
-class WhileStatementWriter implements ASTNodeWriter {
-    write(node: WhileStatement, writer: ASTWriter, fragments: Map<ASTNode, string>): string {
-        const condition = writer.write(node.vCondition, fragments);
-        const body = writer.write(node.vBody, fragments);
-
-        return "while (" + condition + ") " + body;
+class ForStatementWriter extends CompoundStatementWriter<ForStatement> {
+    writeInt(node: ForStatement, writer: ASTWriter): SrcDesc {
+        return writer.desc(
+            "for (",
+            ...(node.vInitializationExpression === undefined
+                ? ["; "]
+                : [node.vInitializationExpression, " "]),
+            node.vCondition,
+            "; ",
+            node.vLoopExpression,
```

Comment:
I believer these two are also [optional](https://github.com/ConsenSys/solc-typed-ast/blob/master/test/samples/solidity/statements/for_050.sourced.sol#L82).

Reply:
The `writer.desc()` function supports optional arguments (i.e. undefined and null)

Reply:
This one got a special case just cause I wanted to write "; " instead of "" when its omitted.